File: chip8.c
#include "chip8.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

// CHIP-8 font set (each digit is 5 bytes)
const uint8_t fontset[80] = {
    0xF0, 0x90, 0x90, 0x90, 0xF0, // 0
    0x20, 0x60, 0x20, 0x20, 0x70, // 1
    0xF0, 0x10, 0xF0, 0x80, 0xF0, // 2
    0xF0, 0x10, 0xF0, 0x10, 0xF0, // 3
    0x90, 0x90, 0xF0, 0x10, 0x10, // 4
    0xF0, 0x80, 0xF0, 0x10, 0xF0, // 5
    0xF0, 0x80, 0xF0, 0x90, 0xF0, // 6
    0xF0, 0x10, 0x20, 0x40, 0x40, // 7
    0xF0, 0x90, 0xF0, 0x90, 0xF0, // 8
    0xF0, 0x90, 0xF0, 0x10, 0xF0, // 9
    0xF0, 0x90, 0xF0, 0x90, 0x90, // A
    0xE0, 0x90, 0xE0, 0x90, 0xE0, // B
    0xF0, 0x80, 0x80, 0x80, 0xF0, // C
    0xE0, 0x90, 0x90, 0x90, 0xE0, // D
    0xF0, 0x80, 0xF0, 0x80, 0xF0, // E
    0xF0, 0x80, 0xF0, 0x80, 0x80  // F
};

void chip8_init(Chip8 *chip8) {
    // Initialize all memory to 0
    memset(chip8, 0, sizeof(Chip8));
    
    // Load fontset into memory [citation:1]
    for (int i = 0; i < 80; i++) {
        chip8->memory[FONT_START + i] = fontset[i];
    }
    
    // Set program counter to start [citation:1]
    chip8->PC = PROGRAM_START;
    
    // Initialize random number generator
    srand(time(NULL));
}

void chip8_reset(Chip8 *chip8) {
    // Clear registers and memory (except font)
    memset(chip8->V, 0, NUM_REGISTERS);
    memset(chip8->stack, 0, STACK_SIZE * sizeof(uint16_t));
    memset(chip8->display, 0, DISPLAY_SIZE * sizeof(uint32_t));
    memset(chip8->keypad, 0, 16);
    
    chip8->I = 0;
    chip8->PC = PROGRAM_START;
    chip8->SP = 0;
    chip8->delay_timer = 0;
    chip8->sound_timer = 0;
    chip8->draw_flag = false;
}

void chip8_load_rom(Chip8 *chip8, const char *rom_path) {
    FILE *rom = fopen(rom_path, "rb");
    if (!rom) {
        printf("Error: Could not open ROM file %s\n", rom_path);
        exit(1);
    }
    
    // Get file size
    fseek(rom, 0, SEEK_END);
    long rom_size = ftell(rom);
    fseek(rom, 0, SEEK_SET);
    
    // Check if ROM fits in memory
    if (rom_size > MEMORY_SIZE - PROGRAM_START) {
        printf("Error: ROM too large for memory\n");
        fclose(rom);
        exit(1);
    }
    
    // Load ROM into memory starting at 0x200 [citation:8]
    fread(&chip8->memory[PROGRAM_START], 1, rom_size, rom);
    fclose(rom);
    
    printf("Loaded ROM: %s (%ld bytes)\n", rom_path, rom_size);
}

uint16_t chip8_fetch_opcode(Chip8 *chip8) {
    // Fetch 2-byte opcode [citation:8]
    return (chip8->memory[chip8->PC] << 8) | chip8->memory[chip8->PC + 1];
}

void chip8_execute(Chip8 *chip8, uint16_t opcode) {
    uint8_t x = (opcode & 0x0F00) >> 8;
    uint8_t y = (opcode & 0x00F0) >> 4;
    uint8_t n = opcode & 0x000F;
    uint8_t kk = opcode & 0x00FF;
    uint16_t nnn = opcode & 0x0FFF;
    
    // Default: increment PC by 2 (will be overridden for jumps/calls)
    int pc_increment = 2;
    
    switch (opcode & 0xF000) {
        case 0x0000:
            switch (opcode & 0x00FF) {
                case 0x00E0: // CLS - Clear display [citation:8]
                    memset(chip8->display, 0, DISPLAY_SIZE * sizeof(uint32_t));
                    chip8->draw_flag = true;
                    break;
                    
                case 0x00EE: // RET - Return from subroutine [citation:8]
                    chip8->SP--;
                    chip8->PC = chip8->stack[chip8->SP];
                    pc_increment = 0;
                    break;
                    
                default: // 0NNN - RCA 1802 program call (ignored)
                    break;
            }
            break;
            
        case 0x1000: // 1NNN - Jump to NNN [citation:8]
            chip8->PC = nnn;
            pc_increment = 0;
            break;
            
        case 0x2000: // 2NNN - Call subroutine at NNN [citation:8]
            chip8->stack[chip8->SP] = chip8->PC + 2;
            chip8->SP++;
            chip8->PC = nnn;
            pc_increment = 0;
            break;
            
        case 0x3000: // 3XNN - Skip next if VX == NN [citation:8]
            if (chip8->V[x] == kk)
                pc_increment = 4;
            break;
            
        case 0x4000: // 4XNN - Skip next if VX != NN [citation:8]
            if (chip8->V[x] != kk)
                pc_increment = 4;
            break;
            
        case 0x5000: // 5XY0 - Skip next if VX == VY [citation:8]
            if (chip8->V[x] == chip8->V[y])
                pc_increment = 4;
            break;
            
        case 0x6000: // 6XNN - Set VX = NN [citation:8]
            chip8->V[x] = kk;
            break;
            
        case 0x7000: // 7XNN - Set VX = VX + NN [citation:8]
            chip8->V[x] += kk;
            break;
            
        case 0x8000:
            switch (n) {
                case 0x0: // 8XY0 - Set VX = VY [citation:8]
                    chip8->V[x] = chip8->V[y];
                    break;
                    
                case 0x1: // 8XY1 - Set VX = VX OR VY [citation:8]
                    chip8->V[x] |= chip8->V[y];
                    break;
                    
                case 0x2: // 8XY2 - Set VX = VX AND VY [citation:8]
                    chip8->V[x] &= chip8->V[y];
                    break;
                    
                case 0x3: // 8XY3 - Set VX = VX XOR VY [citation:8]
                    chip8->V[x] ^= chip8->V[y];
                    break;
                    
                case 0x4: // 8XY4 - Add VY to VX, set VF on carry [citation:8]
                    {
                        uint16_t sum = chip8->V[x] + chip8->V[y];
                        chip8->V[0xF] = (sum > 0xFF) ? 1 : 0;
                        chip8->V[x] = sum & 0xFF;
                    }
                    break;
                    
                case 0x5: // 8XY5 - Subtract VY from VX, set VF on borrow [citation:8]
                    chip8->V[0xF] = (chip8->V[x] >= chip8->V[y]) ? 1 : 0;
                    chip8->V[x] -= chip8->V[y];
                    break;
                    
                case 0x6: // 8XY6 - Shift right VX, set VF to LSB [citation:8]
                    chip8->V[0xF] = chip8->V[x] & 0x1;
                    chip8->V[x] >>= 1;
                    break;
                    
                case 0x7: // 8XY7 - Set VX = VY - VX, set VF on borrow [citation:8]
                    chip8->V[0xF] = (chip8->V[y] >= chip8->V[x]) ? 1 : 0;
                    chip8->V[x] = chip8->V[y] - chip8->V[x];
                    break;
                    
                case 0xE: // 8XYE - Shift left VX, set VF to MSB [citation:8]
                    chip8->V[0xF] = (chip8->V[x] & 0x80) >> 7;
                    chip8->V[x] <<= 1;
                    break;
            }
            break;
            
        case 0x9000: // 9XY0 - Skip next if VX != VY [citation:8]
            if (chip8->V[x] != chip8->V[y])
                pc_increment = 4;
            break;
            
        case 0xA000: // ANNN - Set I = NNN [citation:8]
            chip8->I = nnn;
            break;
            
        case 0xB000: // BNNN - Jump to NNN + V0 [citation:8]
            chip8->PC = nnn + chip8->V[0];
            pc_increment = 0;
            break;
            
        case 0xC000: // CXNN - Set VX = random & NN [citation:8]
            chip8->V[x] = (rand() % 256) & kk;
            break;
            
        case 0xD000: // DXYN - Draw sprite at (VX, VY) with height N [citation:8]
            {
                uint8_t x_pos = chip8->V[x] % DISPLAY_WIDTH;
                uint8_t y_pos = chip8->V[y] % DISPLAY_HEIGHT;
                chip8->V[0xF] = 0;
                
                for (int row = 0; row < n; row++) {
                    uint8_t sprite_byte = chip8->memory[chip8->I + row];
                    
                    for (int col = 0; col < 8; col++) {
                        if (sprite_byte & (0x80 >> col)) {
                            int pixel_x = (x_pos + col) % DISPLAY_WIDTH;
                            int pixel_y = (y_pos + row) % DISPLAY_HEIGHT;
                            int idx = pixel_y * DISPLAY_WIDTH + pixel_x;
                            
                            // Check for collision
                            if (chip8->display[idx] == 0xFFFFFFFF)
                                chip8->V[0xF] = 1;
                            
                            // XOR pixel
                            chip8->display[idx] ^= 0xFFFFFFFF;
                        }
                    }
                }
                chip8->draw_flag = true;
            }
            break;
            
        case 0xE000:
            switch (opcode & 0x00FF) {
                case 0x9E: // EX9E - Skip if key VX pressed
                    if (chip8->keypad[chip8->V[x]])
                        pc_increment = 4;
                    break;
                    
                case 0xA1: // EXA1 - Skip if key VX not pressed
                    if (!chip8->keypad[chip8->V[x]])
                        pc_increment = 4;
                    break;
            }
            break;
            
        case 0xF000:
            switch (opcode & 0x00FF) {
                case 0x07: // FX07 - Set VX = delay timer [citation:8]
                    chip8->V[x] = chip8->delay_timer;
                    break;
                    
                case 0x0A: // FX0A - Wait for key press, store in VX [citation:8]
                    {
                        int key_pressed = 0;
                        for (int i = 0; i < 16; i++) {
                            if (chip8->keypad[i]) {
                                chip8->V[x] = i;
                                key_pressed = 1;
                                break;
                            }
                        }
                        if (!key_pressed)
                            pc_increment = 0; // Repeat instruction
                    }
                    break;
                    
                case 0x15: // FX15 - Set delay timer = VX [citation:8]
                    chip8->delay_timer = chip8->V[x];
                    break;
                    
                case 0x18: // FX18 - Set sound timer = VX [citation:8]
                    chip8->sound_timer = chip8->V[x];
                    break;
                    
                case 0x1E: // FX1E - Set I = I + VX [citation:8]
                    chip8->I += chip8->V[x];
                    break;
                    
                case 0x29: // FX29 - Set I to sprite location for digit VX
                    chip8->I = FONT_START + (chip8->V[x] * 5);
                    break;
                    
                case 0x33: // FX33 - Store BCD of VX in memory [citation:8]
                    chip8->memory[chip8->I] = chip8->V[x] / 100;
                    chip8->memory[chip8->I + 1] = (chip8->V[x] / 10) % 10;
                    chip8->memory[chip8->I + 2] = chip8->V[x] % 10;
                    break;
                    
                case 0x55: // FX55 - Store V0-VX in memory starting at I [citation:8]
                    for (int i = 0; i <= x; i++)
                        chip8->memory[chip8->I + i] = chip8->V[i];
                    break;
                    
                case 0x65: // FX65 - Read V0-VX from memory starting at I [citation:8]
                    for (int i = 0; i <= x; i++)
                        chip8->V[i] = chip8->memory[chip8->I + i];
                    break;
            }
            break;
    }
    
    chip8->PC += pc_increment;
}

void chip8_update_timers(Chip8 *chip8) {
    // Decrement timers at 60 Hz [citation:1][citation:8]
    if (chip8->delay_timer > 0)
        chip8->delay_timer--;
    
    if (chip8->sound_timer > 0) {
        chip8->sound_timer--;
        // In a full implementation, beep when sound_timer > 0
        if (chip8->sound_timer > 0) {
            // printf("\a"); // Terminal bell - simple sound
        }
    }
}

// Simple terminal display for vscode.dev
void chip8_display_terminal(Chip8 *chip8) {
    printf("\033[H"); // Move cursor to home
    
    for (int y = 0; y < DISPLAY_HEIGHT; y++) {
        for (int x = 0; x < DISPLAY_WIDTH; x++) {
            if (chip8->display[y * DISPLAY_WIDTH + x])
                printf("█");
            else
                printf(" ");
        }
        printf("\n");
    }
    printf("\n");
}

// Map keyboard input to CHIP-8 keypad [citation:4]
uint8_t chip8_get_key(void) {
    // This is a simplified version - in real use, you'd use termios for raw input
    // For vscode.dev, you'll need to modify based on your needs
    return 0xFF; // No key pressed
}

File: chip8.js
// chip8.js - Complete CHIP-8 Emulator
class Chip8 {
    constructor() {
        // Memory and registers
        this.memory = new Uint8Array(4096);
        this.V = new Uint8Array(16);        // General purpose registers
        this.I = 0;                          // Index register
        this.PC = 0x200;                      // Program counter
        this.stack = new Uint16Array(16);     
        this.SP = 0;
        
        // Timers
        this.delay_timer = 0;
        this.sound_timer = 0;
        
        // Display (64x32 pixels)
        this.display = new Uint8Array(64 * 32);
        this.draw_flag = false;
        
        // Keypad (16 keys)
        this.keypad = new Uint8Array(16);
        
        // Load font
        this.loadFont();
    }
    
    loadFont() {
        // CHIP-8 font set (each digit is 5 bytes)
        const font = [
            0xF0, 0x90, 0x90, 0x90, 0xF0, // 0
            0x20, 0x60, 0x20, 0x20, 0x70, // 1
            0xF0, 0x10, 0xF0, 0x80, 0xF0, // 2
            0xF0, 0x10, 0xF0, 0x10, 0xF0, // 3
            0x90, 0x90, 0xF0, 0x10, 0x10, // 4
            0xF0, 0x80, 0xF0, 0x10, 0xF0, // 5
            0xF0, 0x80, 0xF0, 0x90, 0xF0, // 6
            0xF0, 0x10, 0x20, 0x40, 0x40, // 7
            0xF0, 0x90, 0xF0, 0x90, 0xF0, // 8
            0xF0, 0x90, 0xF0, 0x10, 0xF0, // 9
            0xF0, 0x90, 0xF0, 0x90, 0x90, // A
            0xE0, 0x90, 0xE0, 0x90, 0xE0, // B
            0xF0, 0x80, 0x80, 0x80, 0xF0, // C
            0xE0, 0x90, 0x90, 0x90, 0xE0, // D
            0xF0, 0x80, 0xF0, 0x80, 0xF0, // E
            0xF0, 0x80, 0xF0, 0x80, 0x80  // F
        ];
        
        // Load font into memory starting at 0x50
        for (let i = 0; i < font.length; i++) {
            this.memory[i] = font[i];
        }
    }

    // Add this test method to the Chip8 class
/*testMode() {
    // Simple test pattern - draws a smiley face
    const testPattern = [
        0b01111110,
        0b10000001,
        0b10100101,
        0b10000001,
        0b10100101,
        0b10011001,
        0b10000001,
        0b01111110
    ];
    
    for (let y = 0; y < 8; y++) {
        for (let x = 0; x < 8; x++) {
            if (testPattern[y] & (0x80 >> x)) {
                this.display[(y + 12) * 64 + (x + 28)] = 1;
            }
        }
    }
    this.draw_flag = true;
}*/
    
    loadROM(romData) {
        // Load ROM into memory starting at 0x200
        for (let i = 0; i < romData.length; i++) {
            this.memory[0x200 + i] = romData[i];
        }
    }
    
    emulateCycle() {
        // Fetch opcode (2 bytes)
        const opcode = (this.memory[this.PC] << 8) | this.memory[this.PC + 1];
        let pc_increment = 2;
        
        // Decode and execute
        const firstNibble = (opcode & 0xF000) >> 12;
        
        switch (firstNibble) {
            case 0x0:
                if (opcode === 0x00E0) { // CLS - Clear display
                    this.display.fill(0);
                    this.draw_flag = true;
                } else if (opcode === 0x00EE) { // RET - Return from subroutine
                    this.SP--;
                    this.PC = this.stack[this.SP];
                    pc_increment = 0;
                }
                break;
                
            case 0x1: // 1NNN - Jump
                this.PC = opcode & 0x0FFF;
                pc_increment = 0;
                break;
                
            case 0x2: // 2NNN - Call subroutine
                this.stack[this.SP] = this.PC + 2;
                this.SP++;
                this.PC = opcode & 0x0FFF;
                pc_increment = 0;
                break;
                
            case 0x3: { // 3XNN - Skip if VX == NN
                const x = (opcode & 0x0F00) >> 8;
                const val = opcode & 0x00FF;
                if (this.V[x] === val) pc_increment = 4;
                break;
            }
            
            case 0x4: { // 4XNN - Skip if VX != NN
                const x = (opcode & 0x0F00) >> 8;
                const val = opcode & 0x00FF;
                if (this.V[x] !== val) pc_increment = 4;
                break;
            }
            
            case 0x5: { // 5XY0 - Skip if VX == VY
                const x = (opcode & 0x0F00) >> 8;
                const y = (opcode & 0x00F0) >> 4;
                if (this.V[x] === this.V[y]) pc_increment = 4;
                break;
            }
            
            case 0x6: { // 6XNN - Set VX = NN
                const x = (opcode & 0x0F00) >> 8;
                this.V[x] = opcode & 0x00FF;
                break;
            }
            
            case 0x7: { // 7XNN - Set VX = VX + NN
                const x = (opcode & 0x0F00) >> 8;
                this.V[x] += opcode & 0x00FF;
                break;
            }
            
            case 0x8: {
                const x = (opcode & 0x0F00) >> 8;
                const y = (opcode & 0x00F0) >> 4;
                const n = opcode & 0x000F;
                
                switch (n) {
                    case 0x0: // 8XY0 - Set VX = VY
                        this.V[x] = this.V[y];
                        break;
                    case 0x1: // 8XY1 - VX = VX OR VY
                        this.V[x] |= this.V[y];
                        break;
                    case 0x2: // 8XY2 - VX = VX AND VY
                        this.V[x] &= this.V[y];
                        break;
                    case 0x3: // 8XY3 - VX = VX XOR VY
                        this.V[x] ^= this.V[y];
                        break;
                    case 0x4: // 8XY4 - Add with carry
                        const sum = this.V[x] + this.V[y];
                        this.V[0xF] = (sum > 0xFF) ? 1 : 0;
                        this.V[x] = sum & 0xFF;
                        break;
                    case 0x5: // 8XY5 - Subtract with borrow
                        this.V[0xF] = (this.V[x] >= this.V[y]) ? 1 : 0;
                        this.V[x] -= this.V[y];
                        break;
                    case 0x6: // 8XY6 - Shift right
                        this.V[0xF] = this.V[x] & 0x1;
                        this.V[x] >>= 1;
                        break;
                    case 0x7: // 8XY7 - Set VX = VY - VX
                        this.V[0xF] = (this.V[y] >= this.V[x]) ? 1 : 0;
                        this.V[x] = this.V[y] - this.V[x];
                        break;
                    case 0xE: // 8XYE - Shift left
                        this.V[0xF] = (this.V[x] & 0x80) >> 7;
                        this.V[x] <<= 1;
                        break;
                }
                break;
            }
            
            case 0x9: { // 9XY0 - Skip if VX != VY
                const x = (opcode & 0x0F00) >> 8;
                const y = (opcode & 0x00F0) >> 4;
                if (this.V[x] !== this.V[y]) pc_increment = 4;
                break;
            }
            
            case 0xA: // ANNN - Set I = NNN
                this.I = opcode & 0x0FFF;
                break;
                
            case 0xB: // BNNN - Jump to NNN + V0
                this.PC = (opcode & 0x0FFF) + this.V[0];
                pc_increment = 0;
                break;
                
            case 0xC: { // CXNN - Random
                const x = (opcode & 0x0F00) >> 8;
                this.V[x] = Math.floor(Math.random() * 256) & (opcode & 0x00FF);
                break;
            }
            
            case 0xD: { // DXYN - Draw sprite
                const x = this.V[(opcode & 0x0F00) >> 8] % 64;
                const y = this.V[(opcode & 0x00F0) >> 4] % 32;
                const height = opcode & 0x000F;
                
                this.V[0xF] = 0;
                
                for (let row = 0; row < height; row++) {
                    const sprite = this.memory[this.I + row];
                    
                    for (let col = 0; col < 8; col++) {
                        if (sprite & (0x80 >> col)) {
                            const pixelX = (x + col) % 64;
                            const pixelY = (y + row) % 32;
                            const idx = pixelY * 64 + pixelX;
                            
                            if (this.display[idx] === 1) {
                                this.V[0xF] = 1;
                            }
                            
                            this.display[idx] ^= 1;
                        }
                    }
                }
                this.draw_flag = true;
                break;
            }
            
            case 0xE: {
                const x = (opcode & 0x0F00) >> 8;
                
                if ((opcode & 0x00FF) === 0x9E) { // EX9E - Skip if key pressed
                    if (this.keypad[this.V[x]]) pc_increment = 4;
                } else if ((opcode & 0x00FF) === 0xA1) { // EXA1 - Skip if key not pressed
                    if (!this.keypad[this.V[x]]) pc_increment = 4;
                }
                break;
            }
            
            case 0xF: {
                const x = (opcode & 0x0F00) >> 8;
                const last = opcode & 0x00FF;
                
                switch (last) {
                    case 0x07: // FX07 - Set VX = delay timer
                        this.V[x] = this.delay_timer;
                        break;
                        
                    case 0x0A: { // FX0A - Wait for key press
                        let pressed = false;
                        for (let i = 0; i < 16; i++) {
                            if (this.keypad[i]) {
                                this.V[x] = i;
                                pressed = true;
                                break;
                            }
                        }
                        if (!pressed) pc_increment = 0; // Repeat instruction
                        break;
                    }
                    
                    case 0x15: // FX15 - Set delay timer
                        this.delay_timer = this.V[x];
                        break;
                        
                    case 0x18: // FX18 - Set sound timer
                        this.sound_timer = this.V[x];
                        break;
                        
                    case 0x1E: // FX1E - Add to I
                        this.I += this.V[x];
                        break;
                        
                    case 0x29: // FX29 - Set I to font character
                        this.I = this.V[x] * 5;
                        break;
                        
                    case 0x33: // FX33 - Store BCD
                        this.memory[this.I] = Math.floor(this.V[x] / 100);
                        this.memory[this.I + 1] = Math.floor((this.V[x] % 100) / 10);
                        this.memory[this.I + 2] = this.V[x] % 10;
                        break;
                        
                    case 0x55: // FX55 - Store registers
                        for (let i = 0; i <= x; i++) {
                            this.memory[this.I + i] = this.V[i];
                        }
                        break;
                        
                    case 0x65: // FX65 - Load registers
                        for (let i = 0; i <= x; i++) {
                            this.V[i] = this.memory[this.I + i];
                        }
                        break;
                }
                break;
            }
        }
        
        this.PC += pc_increment;
        
        // Update timers
        if (this.delay_timer > 0) this.delay_timer--;
        if (this.sound_timer > 0) this.sound_timer--;
    }
}

// Emulator UI Controller
class Chip8Emulator {
    constructor() {
        this.chip8 = new Chip8();
        this.scale = 10;
        this.running = false;
        this.animationFrame = null;
        
        // Set up canvas
        this.canvas = document.getElementById('display');
        this.ctx = this.canvas.getContext('2d');
        this.canvas.width = 64 * this.scale;
        this.canvas.height = 32 * this.scale;
        
        // Key mapping [citation:3]
        this.keyMap = {
            '1': 0x1, '2': 0x2, '3': 0x3, '4': 0xC,
            'q': 0x4, 'w': 0x5, 'e': 0x6, 'r': 0xD,
            'a': 0x7, 's': 0x8, 'd': 0x9, 'f': 0xE,
            'z': 0xA, 'x': 0x0, 'c': 0xB, 'v': 0xF
        };
        
        // Bind events
        this.setupEventListeners();
    }
    
    setupEventListeners() {
        // Keyboard input
        document.addEventListener('keydown', (e) => {
            if (e.key in this.keyMap) {
                this.chip8.keypad[this.keyMap[e.key]] = 1;
                e.preventDefault();
            }
        });
        
        document.addEventListener('keyup', (e) => {
            if (e.key in this.keyMap) {
                this.chip8.keypad[this.keyMap[e.key]] = 0;
                e.preventDefault();
            }
        });
        
        // File upload
        document.getElementById('romFile').addEventListener('change', (e) => {
            this.loadROM(e.target.files[0]);
        });
        
        // Control buttons
        document.getElementById('startBtn').addEventListener('click', () => this.start());
        document.getElementById('pauseBtn').addEventListener('click', () => this.pause());
        document.getElementById('resetBtn').addEventListener('click', () => this.reset());
    }
    
    loadROM(file) {
        const reader = new FileReader();
        reader.onload = (e) => {
            const romData = new Uint8Array(e.target.result);
            this.chip8.loadROM(romData);
            document.getElementById('status').textContent = `Loaded: ${file.name}`;
        };
        reader.readAsArrayBuffer(file);
    }
    
    render() {
        // Clear canvas
        this.ctx.fillStyle = 'black';
        this.ctx.fillRect(0, 0, this.canvas.width, this.canvas.height);
        
        // Draw pixels
        this.ctx.fillStyle = '#00ff00'; // Classic green phosphor
        
        for (let y = 0; y < 32; y++) {
            for (let x = 0; x < 64; x++) {
                if (this.chip8.display[y * 64 + x]) {
                    this.ctx.fillRect(
                        x * this.scale,
                        y * this.scale,
                        this.scale,
                        this.scale
                    );
                }
            }
        }
    }
    
    emulationLoop() {
        if (!this.running) return;
        
        // Run 10 cycles per frame (~600 Hz CPU)
        for (let i = 0; i < 10; i++) {
            this.chip8.emulateCycle();
        }
        
        // Render if needed
        if (this.chip8.draw_flag) {
            this.render();
            this.chip8.draw_flag = false;
        }
        
        // Schedule next frame (~60 FPS)
        this.animationFrame = requestAnimationFrame(() => this.emulationLoop());
    }
    
    start() {
        if (!this.running) {
            this.running = true;
            this.emulationLoop();
            document.getElementById('status').textContent = 'Running...';
        }
    }
    
    pause() {
        this.running = false;
        if (this.animationFrame) {
            cancelAnimationFrame(this.animationFrame);
        }
        document.getElementById('status').textContent = 'Paused';
    }
    
    reset() {
        this.pause();
        this.chip8 = new Chip8();
        /*this.chip8.testMode(); Add this line*/        
        this.render();
        document.getElementById('status').textContent = 'Reset';
    }
}

// Initialize when page loads
window.onload = () => {
    new Chip8Emulator();
};